kubernetes/pods.go

```
@@ -0,0 +1,97 @@
+func (a *Adapter) PodInformer(ctx context.Context, endpointChan chan<- []string) error {
+	podEndpoints := sync.Map{}
+
+	log.Infof("Watching for Pods with labelselector %s in namespace %s", a.cniPodNamespace, a.cniPodLabelSelector)
+	factory := informers.NewSharedInformerFactoryWithOptions(a.clientset, resyncInterval, informers.WithNamespace(a.cniPodNamespace),
+		informers.WithTweakListOptions(func(options *apisv1.ListOptions) { options.LabelSelector = a.cniPodLabelSelector }))
+
+	informer := factory.Core().V1().Pods().Informer()
+	informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		UpdateFunc: func(_, newResource interface{}) {
+			pod, ok := newResource.(*corev1.Pod)
+			if !ok {
+				log.Printf("cannot cast object to corev1.Pod %v", newResource)
+				return
+			}
+			switch {
+
+			case pod.DeletionTimestamp != nil:
+				name, exists := podEndpoints.LoadAndDelete(pod.Name)
+				if !exists {
+					return
+				}
+				log.Infof("Deleted pod: %s IP: %s", pod.Name, name)
+				queueEndpoints(&podEndpoints, endpointChan)
+
+			case pod.Status.ContainerStatuses != nil && len(pod.Status.ContainerStatuses) > 0 &&
+				pod.Status.ContainerStatuses[0].State.Running != nil && pod.Status.PodIP != "":
+				if _, isStored := podEndpoints.LoadOrStore(pod.Name, pod.Status.PodIP); isStored {
+					return
+				}
+				log.Infof("New discovered pod: %s IP: %s", pod.Name, pod.Status.PodIP)
+				queueEndpoints(&podEndpoints, endpointChan)
+			}
+		},
+	})
+
+	factory.Start(ctx.Done())
+	if !cache.WaitForCacheSync(ctx.Done(), informer.HasSynced) {
+		return fmt.Errorf("Timed out waiting for caches to sync")
+	}
+
+	for {
+		list, err := factory.Core().V1().Pods().Lister().List(labels.Everything())
+		if err == nil && len(list) > 0 {
+			break
+		}
+		log.Errorf("Error listing Pods with labelselector %s in namespace %s: %v", a.cniPodNamespace, a.cniPodLabelSelector, err)
+		time.Sleep(resyncInterval)
+	}
```

What is the purpose of this loop? Could you add a comment maybe?

---

Initial test if ingress pods for given specs can be found, otherwise the informer would silently fail

---

Please add this as comment to the code. Sounds like a good decision to check for safety.

---

done